### ISSUE TITLE
Default to square grid for axis('equal')

### ISSUE THREAD

**Issue by Atcold** [matplotlib/matplotlib#17965]
When setting `axis('equal')` I'd expect the grid to be made of squares.
The line *y* = *x* should be bisecting all these squares, and go up 45°.
Instead…

```python
x = linspace(0, 1.5)
y = sin(2 * π * x)
plot(x, y)
axis('equal')
```

![sine-equal](https://user-images.githubusercontent.com/2119355/87866424-cdc41300-c94f-11ea-9ca9-df39d0e9fd64.png)

while `axis('square')` gets the squared grid, *but* changes the axis length (as expected).

```python
plot(x, y)
axis('square')
```

![sine-square](https://user-images.githubusercontent.com/2119355/87866436-fea44800-c94f-11ea-8118-87e67d925109.png)

My expectation would be having a squared grid when issuing `axis('equal')` (which is the reason behind setting a 1:1 aspect ratio).

> P.S. I come to think that *maybe* so many *x*-ticks do not fit. Is there a way to enforce the grid, while skipping every other displayed number, perhaps? I'm looking for defaults, not to solve this specific issue.

> P.P.S. There's something funky with the location of the point `(0, 0)` too…

**Comment by timhoffm** [comment]
The grid is determined by the ticks. `axis('equal')` does not have any control over these. In particular, the ticks are determinded by the locator ([AutoLocator](https://matplotlib.org/api/ticker_api.html#matplotlib.ticker.AutoLocator) by default). - It tries to find a reasonable number of ticks at nice locations for a given range. If the axis is not square, there are different ranges for x and y, which may result in different tick positions for x and y and thus a non-square grid.

In general, "a reasonable number of ticks at nice positions" and "same tick distances on both axis" are contradicting requirements. Additionally, the latter requires a coordination between both Axis, which is not part of the current Matplotlib architecture (Axis are independent of each other).

You have the following options:

- go with square figures
- manually tune the tick positions
- allow more ticks on the x-axis
  ~~~
  aspect = fig.get_figwidth() / fig.get_figheight()
  ax.xaxis.set_major_locator(MaxNLocator(nbins=10*aspect))
  ~~~
- limit the possible locations of ticks
  ~~~
   ax.xaxis.set_major_locator(MaxNLocator(steps=[1, 5]))
   ax.yaxis.set_major_locator(MaxNLocator(steps=[1, 5]))
   ~~~
   This should work for not too elongated plots.

There's nothing we can do about it from the library side.

**Comment by timhoffm** [comment]
The general issue is tracked in #8469.

**Comment by Atcold** [comment]
Wow @timhoffm, thank you for the lengthy explanation and workaround.
I didn't get the

> Axis, which is not part of the current Matplotlib architecture (Axis are independent of each other)

What is independent of what? I also don't really understand the difference in `axis` and `axes`, but I think this is coming from MATLAB.

**Comment by timhoffm** [comment]
Indeed the terminology `Axes` comes from MATLAB. An `Axes` is essentially a (sub)plot, i.e. it contains, the x-axis, y-axis (including labels ticks etc, a title and the plotted data.

![grafik](https://user-images.githubusercontent.com/2836374/89121598-70d46b00-d4c0-11ea-9656-c20c9bc1845f.png)

The x-axis and the y-axis are independed of each other. The y-axis does not know about the x-axis and vice-versa. They individually determine how many and where to put ticks. Thus it's difficult to assure they use the same tick distance to ensure a square grid.

**Comment by Atcold** [comment]
Can't we have `AutoLocator` look up if axes have been set to `'equal'`?

**Comment by dopplershift** [comment]
But there's a separate instance of `AutoLocator` for the x-axis and the y-axis. How are those supposed to communicate their spacings to each other?

**Comment by Atcold** [comment]
```python
if gca().aspect==1:
    xAutoLocator.looksAt(yAutoLocator)
```

**Comment by jklymak** [comment]
If someone can get this to work robustly without major spaghetti throughout the locator code, and the `ax.apply_aspect` code, then I'm sure Matplotlib would appreciate a PR.  But before starting such a project, @timhoffm said it nicely above.  

>In general, "a reasonable number of ticks at nice positions" and "same tick distances on both axis" are contradicting requirements. Additionally, the latter requires a coordination between both Axis, which is not part of the current Matplotlib architecture (Axis are independent of each other).
